x/nexus/keeper/msg_dispatcher.go

```diff
@@ -60,18 +60,17 @@ func (m Messenger) routeMsg(ctx sdk.Context, msg exported.WasmMessage) error {
 		return fmt.Errorf("recipient chain %s is not a registered chain", msg.DestinationChain)
 	}
 
-	id, _, _ := m.GenerateMessageID(ctx)
 	sourceChain := exported.Chain{Name: msg.SourceChain, SupportsForeignAssets: false, KeyType: tss.None, Module: wasmtypes.ModuleName}
 	sender := exported.CrossChainAddress{Chain: sourceChain, Address: msg.SourceAddress}
 	recipient := exported.CrossChainAddress{Chain: destinationChain, Address: msg.DestinationAddress}
 
-	nexusMsg := exported.NewGeneralMessage(id, sender, recipient, msg.PayloadHash, msg.SourceTxID, msg.SourceTxIndex, nil)
-	if err := m.Nexus.SetNewMessage(ctx, nexusMsg); err != nil {
-		return err
-	}
+	nexusMsg := exported.NewGeneralMessage(msg.ID, sender, recipient, msg.PayloadHash, msg.SourceTxID, msg.SourceTxIndex, nil)
 
 	// try routing the message
 	_ = utils.RunCached(ctx, m, func(ctx sdk.Context) (struct{}, error) {
+		if err := m.Nexus.SetNewMessage(ctx, nexusMsg); err != nil {
+			return struct{}{}, err
+		}
 		return struct{}{}, m.RouteMessage(ctx, nexusMsg.ID)
```

Comment:
This works for EVM destinations where `RouteMessage` essentially enqueues the command for signing. Although, I realized this is an issue for Cosmos destinations where the `RouteMessage` will fail since there's no payload attached in a routing context but is needed for the IBC transfer. What was happening before is that `SetNewMessage` would succeed and emit a `MessageReceived` event, but `RouteMessage` would fail for Cosmos. But the relayer would react to the event and still call `axelarnet.RouteMessage` but with the payload attached making it go through. Now `SetNewMessage` will revert as well, so event won't be emitted and relayer won't try this out.

Looks like we should just check if message exists first and return as a no-op
```
	if _, ok := k.GetMessage(ctx, msg.ID); ok {
		return nil
	}
```